### ISSUE TITLE
Failed to fetch dynamically imported module "/_nuxt/OpeningHours.CTqYYKGj.js"

### ISSUE THREAD

**Issue by mikeCrafted** [nuxt/nuxt#26565]
### Environment

------------------------------
- Operating System: Windows_NT
- Node Version:     v21.6.1
- Nuxt Version:     3.11.1
- CLI Version:      3.11.1
- Nitro Version:    2.9.5
- Package Manager:  npm@9.6.4
- Builder:          -
- User Config:      runtimeConfig, sourcemap, app, css, modules, i18n, alias, auth, build, site, schemaOrg, ogImage, sitemap, devtools      
- Runtime Modules:  @sidebase/nuxt-auth@0.6.7, @nuxtjs/i18n@8.2.0, @pinia/nuxt@0.5.1, nuxt3-leaflet@1.0.12, @nuxtjs/seo@2.0.0-rc.10
- Build Modules:    -
------------------------------

### Reproduction

This is an indeterministic bug in production, I cannot find commonalities when it occurs. Apple users (Safari browser) tend to have it more often. I get a lot of sentry errors and see it in the server logs, but mixed with a lot of correct requests, it works on all my devices and on ones of my colleagues, but according to sentry there are users who experience crashes.

### Describe the bug

According to our sentry report, there are regular crashes of the app on many different pages (sometimes on the homepage, sometimes during search etc.), when nuxt fails to resolve a file path and the app crashes. 

The majority of logs are fine, but still we see a regular issues popping up in sentry. The components that fail to resolve are also different. They work on all my devices, so I cannot reproduce the error in production. Nor does it occur in development. 

We host the app on Azure App Service.

I recently upgraded to the latest nuxt version, where the naming of the chunks was changed, still I saw some logs today that a component with the old naming style could not be imported. I am not opting in for the old one, I switched to the default new one.

Can setting the `builder: 'webpack'` help?

I feel like this issue might be related @danielroe ? #21780 

![nuxt-error](https://github.com/nuxt/nuxt/assets/63555352/94b45b55-6fe8-409c-9cca-83e051dc0c46)

**Comment by danielroe** [comment]
I suspect this is due to new deployments and failing to load a dynamic chunk after a new deployment occurs, which might have invalidating existing public assets. We currently handle chunk loading errors when navigating across routes but we could probably think about whether there is anything we could do when that is a dynamic _component_ import.

**Comment by mikeCrafted** [comment]
@danielroe is there anything I can do to avoid this crashing in production? did I understand correctly that this is due to the fact that I use Lazy components - i.e. dynamically loaded ones? Or is this an internal issue?

**Comment by michalzaq12** [comment]
I can confirm that this issue (and others like it) occurs after a new deployment. This isn't just related to "Lazy Components" as it gives me the "Failed to load dynamically..." message for pages, layouts and middleware modules (chunks) as well. My observations when the error occurs:

1. Lazy components
2. The new deployment has not yet been detected by the `check-outdated-build` plugin
3. The new deployment is detected by `check-outdated-build` and handled by Nuxt, but an unhandled exception still occurs (this error should be ignored)
4. After a new deployment, somehow after a new visit, the old version of the page is loaded (how? Browser cache?) and immediately crashes with a "Failed to load dynamically..." for chunks that have changed (new file hash?)

**Comment by mikeCrafted** [comment]
@michalzaq12 thank you for your observations! So I am using GitHub actions with Azure app service to deploy the app. Will a manual build an deploy via file upload for example help?

**Comment by michalzaq12** [comment]
@mikeCrafted it won't help. For now, there is no way to avoid these errors.

**Comment by meirroth** [comment]
We've also encountered these errors after deployments, which we fixed by triggering page reload on chunk error by adding the following file `plugins/error-handler.ts`:
```
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hook('app:chunkError', () => {
    window.location.reload()
  })
})
```
I believe Nuxt should handle this internally, as I am sure it is a common problem.
Related Issues can be found on this thread: https://github.com/nuxt/nuxt/issues/14594

**Comment by mikeCrafted** [comment]
@meirroth I think this handler is already added by default in Nuxt, see #19086 
am I right, @danielroe ? Or can this custom plugin indeed make a difference?

**Comment by meirroth** [comment]
@mikeCrafted From what I understood, and I could be wrong, the built in handler takes care of errors only [during route navigation](https://github.com/nuxt/nuxt/pull/19086/files#diff-1ad1e74ffd33ee9d0f230b707fc20169f67984075c1e977e86cbd2d1870b64ccR54). The plugin I shared above helps with any other chunk loading errors, such as lazy components.

**Comment by mikeCrafted** [comment]
@meirroth oh, that is interesting! why are you not using `reloadNuxtApp()` instead of `window.location.reload()`?

**Comment by michalzaq12** [comment]
> We've also encountered these errors after deployments, which we fixed by triggering page reload on chunk error by adding the following file `plugins/error-handler.ts`:
> 
> ```
> export default defineNuxtPlugin((nuxtApp) => {
>   nuxtApp.hook('app:chunkError', () => {
>     window.location.reload()
>   })
> })
> ```
> 
> I believe Nuxt should handle this internally, as I am sure it is a common problem. Related Issues can be found on this thread: #14594

I see the possibility of an endless reload loop. That's probably why Nuxt doesn't support it internally.

**Comment by mikeCrafted** [comment]
@michalzaq12 adding a query parameter to the URL would help, right? so that we can tell if a reload was already attempted.

**Comment by meirroth** [comment]
> @meirroth oh, that is interesting! why are you not using `reloadNuxtApp()` instead of `window.location.reload()`?

@mikeCrafted  Haven't tried that. I'll give it it a shot now.


> I see the possibility of an endless reload loop. That's probably why Nuxt doesn't support it internally.

@michalzaq12 How so? I haven't encountered any issue in my testing, and got this running in production.

**Comment by meirroth** [comment]
@mikeCrafted `reloadNuxtApp()` [is a wrapper](https://github.com/nuxt/nuxt/blob/17488508b1cf7cfd98462bd85a1dfbf8e0570326/packages/nuxt/src/app/composables/chunk.ts#L59) for `window.location.reload()` with additional features like ttl and persistent useState. So considering we use Pinia and not useState, and I didn't need ttl `window.location.reload()` was good on it's own. But they both work, I just tested.

**Comment by meirroth** [comment]
@mikeCrafted [torahlive.com](https://torahlive.com/) 😃

**Comment by michalzaq12** [comment]
To avoid most errors:
1. Do not delete old files during deployment, allow the user to smoothly update to the new version after a new visit (or when `check-outdated-build` detects new version)
2. Make sure that all .html files are not cached (check `Cache-Control` header)
3. Finally, you can add a plugin for users who still have a problem:
```js
export default defineNuxtPlugin(nuxtApp => {
  const MAX_RETRIES = 2;
  const QUERY_PARAM_NAME = 'nuxt_reload_attempt'

  // Handle "Failed to fetch dynamically imported module ..." or similar issues
  nuxtApp.hook('app:chunkError', () => {
    const searchParams = new URLSearchParams(window.location.search);
    const currentRetry = Number(searchParams.get(QUERY_PARAM_NAME)) || 0
    if (currentRetry < MAX_RETRIES) {
      console.log('[nuxt]: Reloading due to chunk error')
      searchParams.set(QUERY_PARAM_NAME, (currentRetry + 1).toString())
      // Changing the search also causes a refresh
      window.location.search = searchParams.toString()
    }
  })
})
```

**Comment by meirroth** [comment]
> I see the possibility of an endless reload loop. That's probably why Nuxt doesn't support it internally.

@michalzaq12 How so? I haven't encountered any issue in my testing, and got this running in production.

**Comment by nandi95** [comment]
Duplicate of https://github.com/nuxt/nuxt/issues/23612

**Comment by wgcorrea** [comment]
> To avoid most errors:
> 
> 1. Do not delete old files during deployment, allow the user to smoothly update to the new version after a new visit (or when `check-outdated-build` detects new version)
> 2. Make sure that all .html files are not cached (check `Cache-Control` header)
> 3. Finally, you can add a plugin for users who still have a problem:
> 
> ```js
> export default defineNuxtPlugin(nuxtApp => {
>   const MAX_RETRIES = 2;
>   const QUERY_PARAM_NAME = 'nuxt_reload_attempt'
> 
>   // Handle "Failed to fetch dynamically imported module ..." or similar issues
>   nuxtApp.hook('app:chunkError', () => {
>     const searchParams = new URLSearchParams(window.location.search);
>     const currentRetry = Number(searchParams.get(QUERY_PARAM_NAME)) || 0
>     if (currentRetry < MAX_RETRIES) {
>       console.log('[nuxt]: Reloading due to chunk error')
>       searchParams.set(QUERY_PARAM_NAME, (currentRetry + 1).toString())
>       // Changing the search also causes a refresh
>       window.location.search = searchParams.toString()
>     }
>   })
> })
> ```


Doc says that hard reload at the route level is done by default: https://nuxt.com/docs/guide/going-further/experimental-features#emitroutechunkerror. 

If the idea is to watch for that hook and refresh the window, I believe adjusting the `emitRouteChunkError` to `manual` might be required to avoid weird behaviour or making hard to debug what really is fixing the issue.

I'm having the same issue. Currently, deploying at Cloudflare Pages; and all cache settings were reviewed. We tried using a similar approach as the proposed plugin; even adding a `t=Date.now()` to the URL upon refresh to double down in the cache refresh force was not enough.

**Comment by 1cor13** [comment]
> I've just started to have these errors on production, after updating to v3.12.2.
> 
> Solved it by replacing
> 
> ```
> experimental: {
>     componentIslands: {
>         selectiveClient: 'deep'
>     }
> }
> ```
> 
> with
> 
> ```
> experimental: {
>     componentIslands: 'local+remote'
> }
> ```
> 
> as per docs https://nuxt.com/docs/api/components/nuxt-island.

@urbgimtam Do you have any component called 'nuxt-island' in your code base?

**Comment by urbgimtam** [comment]
No, this is configuration on nuxt.config.ts.

A quarta, 19/06/2024, 15:20, 1cor13 ***@***.***> escreveu:

**Comment by DbPtZn** [comment]
I also encountered this problem when I used nginx and a domain name for reverse proxy. The error did not occur when I accessed the port directly using the IP address. The problem was that it was normal once in the middle, and I thought I had solved the problem until I updated the files the next time.

**Comment by awacode21** [comment]
Upstream issue in vite: 
https://github.com/vitejs/vite/issues/11804

**Comment by leopoldkristjansson** [comment]
> We've also encountered these errors after deployments, which we fixed by triggering page reload on chunk error by adding the following file `plugins/error-handler.ts`:
> 
> ```
> export default defineNuxtPlugin((nuxtApp) => {
>   nuxtApp.hook('app:chunkError', () => {
>     window.location.reload()
>   })
> })
> ```
> 
> I believe Nuxt should handle this internally, as I am sure it is a common problem. Related Issues can be found on this thread: #14594

Thanks for sharing this. Did you stop seeing the errors in Sentry after adding this? Or is this just making sure the app reloads if an error of this kind comes up?

**Comment by meirroth** [comment]
@leopoldkristjansson Sure! It didn’t stop the errors in Sentry, but the issue is fixed from the user’s perspective.

**Comment by leopoldkristjansson** [comment]
> @leopoldkristjansson Sure! It didn’t stop the errors in Sentry, but the issue is fixed from the user’s perspective.

@meirroth - OK, exactly. Thanks again!

**Comment by danielroe** [comment]
The next release of Nuxt v3 will ship a significant fix to this - hopefully you might be able to avoid custom solutions then 🙏  (Let me know if that doesn't resolve things for you!)

**Comment by BlazingTide** [comment]
Hey, is there any update for this? 

Also, is it possible to somehow store these chunks on a CDN to prevent new deployments from breaking existing users' apps?

**Comment by 1cor13** [comment]
> The next release of Nuxt v3 will ship a significant fix to this - hopefully you might be able to avoid custom solutions then 🙏 (Let me know if that doesn't resolve things for you!)

@danielroe which version has this issue fixed?

**Comment by MarkusLund** [comment]
Any updates on this?

**Comment by CodeDredd** [comment]
@BlazingTide @1cor13 @MarkusLund Here is the fix from @danielroe 
https://github.com/nuxt/nuxt/commit/820908696

Released mit der nuxt version 3.13.2

**Comment by luizzappa** [comment]
> [@BlazingTide](https://github.com/BlazingTide) [@1cor13](https://github.com/1cor13) [@MarkusLund](https://github.com/MarkusLund) Here is the fix from [@danielroe](https://github.com/danielroe) [820908696](https://github.com/nuxt/nuxt/commit/820908696)
> 
> Released with nuxt version 3.13.2


The problem continued here even after updating to this version or using any of the mentioned plugins. The problem is when you press the browser's back button.

**Comment by TonhaoSantos** [comment]
I managed to get around the error by following the tips suggested here. Thanks @meirroth 

Here is my plugin "0.error-handler.client.ts" with three ways to get around this localhost error:

```
export default defineNuxtPlugin(() => {
  const router = useRouter()

  if (process.env.NODE_ENV === 'development') {
    // Mode 1
    router.onError((error) => {
      if (error.message.includes('Failed to fetch dynamically imported module')) {
        window.location.reload()
      }
    })

    /*
    // Mode 2
    context.hook('app:chunkError', () => {
      window.location.reload()
    })
    */

    /*
    // Mode 3
    context.hook('app:error', (error) => {
      if (error.message.includes('Loading chunk')) {
        window.location.reload()
      }
    })
    */
  }
})
```

**Comment by Chippd** [comment]
I'm still seeing this on 3.14.1592 - it's triggering the full error page I've defined as described in the documentation [here](https://nuxt.com/docs/getting-started/error-handling#error-page) .

From my understanding of the error handling in [nuxt.ts](https://github.com/nuxt/nuxt/blob/82090869645dc37a770c58cc77eb60a02a2e683c/packages/nuxt/src/app/nuxt.ts#L381), it's checking for CSS failures, but not any JS ones (hence the need for your plugin above @TonhaoSantos (which I'm trying out now).

Seems like an easy one to add on the face of it, I'd be happy to try submit a pull request to add that error check, would save us needing plugins and custom solutions

**Comment by hubfactory** [comment]
I've seen a similar issue with Sentry. I think the only solution is to do an Atomic Deployment on the infrastructure side and leave the old version of the file for a certain period of time.

**Comment by Chippd** [comment]
Sharing my latest iteration of this after encountering this error in staging again (on `"nuxt": "^3.14.1592"`)



```
export default defineNuxtPlugin(() => {
  const router = useRouter()

  // Track reload attempts using module-level variables
  const MAX_RELOAD_ATTEMPTS = 3
  const RELOAD_ATTEMPT_TIMEOUT = 30000 // 30 seconds
  let reloadAttempts = 0
  let lastAttemptTime = 0
  
  router.onError((error) => {
    if (
      error.message.includes('Failed to fetch dynamically imported module') ||
      error.message.includes('Loading chunk') ||
      error.message.includes('Failed to load resource')
    ) {
      console.error('Chunk loading error detected:', error)
      
      const now = Date.now()
      
      // Reset count if more than 30 seconds have passed since last attempt
      if (now - lastAttemptTime > RELOAD_ATTEMPT_TIMEOUT) {
        reloadAttempts = 0
        lastAttemptTime = now  // Reset the timestamp when we reset the count
      }
      
      // Only reload if we haven't exceeded max attempts
      if (reloadAttempts < MAX_RELOAD_ATTEMPTS) {
        reloadAttempts++
        
        // Attempt reload with cache-busting parameter
        setTimeout(() => {
          window.location.href = window.location.href + 
            (window.location.href.includes('?') ? '&' : '?') + 
            '_t=' + now
        }, 1000)
      } else {
        console.error('Max reload attempts reached. Please try clearing your browser cache or contact support.')
        // Optionally show user-friendly error message or redirect to error page
      }
    }
  })
})
```

**Comment by emptypockts** [comment]
hi all, 
i had the same issue and the problem was that my app had this cache config wrong 
cache-control public, max-age=31536000, immutable.
i am using cloudflared for my app so i assume that the problem was that my cache was being utilized even after building a new release. i purged my cache in cloudflared and then i updated my app settings with the lines below. 

  routeRules:{
    '/_nuxt/**':{
      headers:{
        'Cache-Control': 'public, max-age=0, must-revalidate',
      }

this fixed the issue for me. hope it helps

**Comment by haexhub** [comment]
i have the same problem when i try to deploy a tauri application to android. the “reload hack” ends in an endless loop. on desktop build however, everything is working fine. Not sure if this is a tauri or nuxt problem.

**Comment by Tofandel** [comment]
This is still an issue
https://github.com/nuxt/nuxt/issues/26565#issuecomment-2541259485
Is spot on, so the reload plugin is somehow still necessary

Are we sure that the router error is not triggering before the app:chunkError?

https://github.com/nuxt/nuxt/commit/820908696#diff-04bd18e1bfe9c11c2c833477d4de246b7e0b13d4f1d721b1a540a9f3c98ea412L29

**Comment by yulafezmesi** [comment]
some of users are having same issue with using Cloudflare cache. after disabling cloudflare proxy seems solve issue. not sure how i can replicate the issue.

**Comment by simonmaass** [comment]
i am also having this issue on deployments!

**Comment by emptypockts** [comment]
@simonmaass this is what worked for me.

**Comment by vinoddotcom** [comment]
i am also having this issue on deployments!

**Comment by neetu1123** [comment]
I am also having this issue on deployments!

**Comment by NormVg** [comment]
samee here 
, btw i am only having this error in the firefox browser only

**Comment by milkyfuckinway** [comment]
We have been encountering this error for several months now, despite the fact that we do not use dynamic import or SSR, only CSR. The error appears mainly on iOS, iPadOS, and macOS, although it sometimes appears in other browsers as well, but less frequently. After a new deployment, the application may become unusable, and it is unclear how to fix this. The workaround with window.location.reload() is basically something beyond the pale.

**Comment by sewalsh** [comment]
I strongly suspect this is due to either an absent `Cache-Control` header or one exists but with a long `max-age` value.

It would be helpful if those having issues include the `Cache-Control` headers used for the `.html` files, they should always be revalidated btw:

`'Cache-Control': 'public, max-age=0, must-revalidate'`

**Comment by emptypockts** [comment]
> hi all, 
> i had the same issue and the problem was that my app had this cache config wrong 
> cache-control public, max-age=31536000, immutable.
> i am using cloudflared for my app so i assume that the problem was that my cache was being utilized even after building a new release. i purged my cache in cloudflared and then i updated my app settings with the lines below. 
> 
>   routeRules:{
>     '/_nuxt/**':{
>       headers:{
>         'Cache-Control': 'public, max-age=0, must-revalidate',
>       }
> 
> this fixed the issue for me. hope it helps

@sewalsh this is how i fixed this issue.

**Comment by hymair** [comment]
What are the performance implications with this approach?

```ts
routeRules: {
    '/_nuxt/**': {
        headers: {
            'Cache-Control': 'public, max-age=0, must-revalidate',
        },
    },
},
```

**Comment by Tofandel** [comment]
It's terrible and the incorrect approach, this would result in no caching for css files, js files and images in the nuxt folder. Those are already cache busted via a hash in the filename. Only the main request should get this header to serve a non cached html.

**Comment by hymair** [comment]
Yeah, thought the same. 

If new deploy happens while users are active, they would still get this error as soon as the deploy is done and old files are cleared, any further navigation would be pointing to a missing file. In this case no request is made for html, so not caching it won't help.

In the case of OP it is a component. So as soon as you navigate to a page using that component, users will get this error. What if Nuxt does a page reload for components as well, same as it does for routes?

**Comment by emptypockts** [comment]
sorry it did that work for you guys. it was not a terrible approach for me. my app works 100%  fine so please ask before stating this is a terrible approach.

**Comment by emptypockts** [comment]
> What are the performance implications with this approach?
> 
> ```ts
> routeRules: {
>     '/_nuxt/**': {
>         headers: {
>             'Cache-Control': 'public, max-age=0, must-revalidate',
>         },
>     },
> },
> ```

hello.
my application has a limited audience. max 15 users and it has no css or any images to load. i am no expert but this may not be the best approach for heavily used apps where performance is prioritized over reliability. this work around fixed my problem and my 15 users have no reported issues since then.

**Comment by sovietsolider** [comment]
Same issue, when deploying via CF. GoogleBot crawled 500 Error after deploy, but re-indexing in Search Console fix the issue.

**Comment by carlosvaldesweb** [comment]
I have the same problem, and it started happening after updating to Nuxt version 3.17. The issue sometimes seems to be fixed for users by using the Nuxt hook they recommend:

```js
nuxtApp.hook("app:chunkError", () => {
  // ...
});
```
However, this doesn't work for the Google crawler. This is killing our SEO—we've been unable to fix it for over two weeks.

Google shows "Failed to fetch dynamically" messages, and I’ve created an issue for this in case someone else is experiencing the same problem and wants more information:

https://github.com/nuxt/nuxt/issues/32795

<img width="1344" height="701" alt="Image" src="https://github.com/user-attachments/assets/33631f7a-118f-4c38-8e23-072aba5aee78" />

<img width="772" height="164" alt="Image" src="https://github.com/user-attachments/assets/82901902-ef21-40ba-89ea-da3acb7593dd" />

The only solution I’ve found so far is to keep re-indexing constantly in Google Search Console. This didn’t happen before updating.


I saw that a fix was made a few months ago for something similar, but it doesn't seem to be working and is affecting many users:
https://github.com/nuxt/nuxt/pull/32137

Do you have any information that could help us?

@TheAlexLichter @danielroe

**Comment by sewalsh** [comment]
@carlosvaldesweb Click on the 'More Info' tab to inspect the HTTP headers. Ensure there's no caching on .html pages.

<img width="1222" height="646" alt="Image" src="https://github.com/user-attachments/assets/5d458b5a-7c26-4150-b1c3-e1ba0ea9f917" />

**Comment by bpcs-jan** [comment]
Any updates on this? This is killing our SEO ranking.

**Comment by danielroe** [comment]
the issue with seo has been resolved if you are on the latest nuxt + rollup versions

it may take some time for Google to update your listings

**Comment by sewalsh** [comment]
@danielroe I think this is a chicken and egg problem. Chunk errors are being caused by the absence of cache directives on html pages, this will be especially noticeable if you push to prod frequently.

I would suspect Vercel users with skew protection haven't hit against this problem nor Netlify users as the nitro preset automatically adds a `must-revalidate`. https://github.com/nitrojs/nitro/blob/bdd3845a75ee884b645705751ea5070cd10b8b31/src/presets/netlify/runtime/netlify.ts#L18 Doing a search of `must-validate` in the Nitro repo yields no other results.

Because nitro/nuxt uses radix3 for pattern matching, it's not possible to match against `*.html`, you'll need to 'exclude' routes by specifying their cache length and then apply a no-cache rule for everything else.

To cache bust your HTML pages, these routeRules will achieve that. (if you have custom asset or public directories, make sure to exclude those).

``` ts
  routeRules: {
    // 1️⃣ Static assets: keep for a year
    '/_nuxt/**': {
      headers: { 'cache-control': 'public, max-age=31536000, immutable' },
    },
    // any other paths
    '/assets/**': {
      headers: { 'cache-control': 'public, max-age=31536000, immutable' },
    },

    // 2️⃣ “All other routes” → these are your HTML pages
    '/**': {
      headers: { 'cache-control': 'public, max-age=0, must-revalidate' },
    },
  },
```
Updating the other non-netlify nitro presets with a default fallback of `"Cache-Control": "public, max-age=0, must-revalidate"` makes most sense to me as way of resolving this.

**Comment by danielroe** [comment]
thoughts @pi0?

**Comment by jlemonz** [comment]
> the issue with seo has been resolved if you are on the latest nuxt + rollup versions
> 
> it may take some time for Google to update your listings

How do i update rollup version? Sorry for the noob question

**Comment by carlosvaldesweb** [comment]
> > the issue with seo has been resolved if you are on the latest nuxt + rollup versions
> > it may take some time for Google to update your listings
> 
> How do i update rollup version? Sorry for the noob question

Run to show your rollup version
```
npm list rollup
```
Check the last version in rollup github
<img width="356" height="542" alt="Image" src="https://github.com/user-attachments/assets/8fdd4ec4-4cf7-4445-8d6c-1d182ac625f7" />

And listall the last version with
```
npm install rollup@4.46
```
Or change it directly in your packeg json and run npm installa after that

**Comment by carlosvaldesweb** [comment]
> [@carlosvaldesweb](https://github.com/carlosvaldesweb) Click on the 'More Info' tab to inspect the HTTP headers. Ensure there's no caching on .html pages.
> 


@sewalsh thanks, These are my headers — do you think I should modify them?

<img width="589" height="625" alt="Image" src="https://github.com/user-attachments/assets/b2100c69-9845-47b2-8791-7ee697848692" />

And @danielroe, I’ve updated to Nuxt 3.18 and Rollup version 4.46.2, but the SEO issue is still happening.

**Comment by carlosvaldesweb** [comment]
Some Nuxt Pages has the error.

![Image](https://github.com/user-attachments/assets/1b972a3b-0149-4290-a670-a713bac6aa47)
![Image](https://github.com/user-attachments/assets/414b9379-f770-4031-ac69-4add41ee2eaa)
![Image](https://github.com/user-attachments/assets/ad305a9d-e3bf-4a87-8b34-902b7a42e9d7)

**Comment by lixiao258** [comment]
> 某些 Nuxt 页面出现错误。
> 
> ![Image](https://github.com/user-attachments/assets/1b972a3b-0149-4290-a670-a713bac6aa47) ![Image](https://github.com/user-attachments/assets/414b9379-f770-4031-ac69-4add41ee2eaa) ![Image](https://github.com/user-attachments/assets/ad305a9d-e3bf-4a87-8b34-902b7a42e9d7)

<img width="1366" height="28" alt="Image" src="https://github.com/user-attachments/assets/a53afd38-e27f-4f75-bd0b-53da38404772" />The same error persists and cannot be resolved, occasionally occurring

**Comment by remihuigen** [comment]
> Updating the other non-netlify nitro presets with a default fallback of `"Cache-Control": "public, max-age=0, must-revalidate"` makes most sense to me as way of resolving this.

@sewalsh I don't agree that just disabling cache for page routes is a good way of resolving this. What's wrong with just making the static assets immutable though? 🤔 
```
'/_nuxt/**': {
      headers: { 'cache-control': 'public, max-age=31536000, immutable' },
}
```

Any user that still has a cached page can request the old assets from cache, while users with 'fresh' html pages will request the new asset.

That, combined with a cache flush upon new deployments of any html that is cached on the server, seems like a better solution to me, where the core issue is resolved, with minimal performance impact

**Comment by boussadjra** [comment]
I faced the same issue and I found that I'm using assets (images) from the public folder like `<v-img src="/images/logo.png"`

**Comment by toniengelhardt** [comment]
Seeing this also with the latest Nuxt 4.

<img width="699" height="555" alt="Image" src="https://github.com/user-attachments/assets/0243839d-b471-47d2-bc00-8b8cea5c7b93" />

**Comment by joeczucha** [comment]
@toniengelhardt 

We launched our new site at the end of July, and by mid-August [our SERPs](https://www.google.com/search?q=500+-+Failed+to+fetch+dynamically&sca_esv=bd4826c8ba41a3de&biw=2560&bih=1278&tbm=nws&sxsrf=AE3TifPmyo7BdRo_4gYR_h2zzQQP_W8RFw%3A1756241126924&ei=5hyuaJGaOMrXhbIP6tbakQY&ved=0ahUKEwiRpYqErKmPAxXKa0EAHWqrNmIQ4dUDCA4&uact=5&oq=500+-+Failed+to+fetch+dynamically&gs_lp=Egxnd3Mtd2l6LW5ld3MiITUwMCAtIEZhaWxlZCB0byBmZXRjaCBkeW5hbWljYWxseTIGEAAYFhgeMgYQABgWGB4yBhAAGBYYHjILEAAYgAQYhgMYigUyCxAAGIAEGIYDGIoFMgsQABiABBiGAxiKBTIIEAAYgAQYogQyBRAAGO8FMggQABiiBBiJBTIIEAAYgAQYogRIyAVQAFgAcAB4AJABAJgBWKABWKoBATG4AQPIAQCYAgGgAmKYAwCIBgGSBwExoAePB7IHATG4B2LCBwMyLTHIBwc&sclient=gws-wiz-news) were full of “500 – Failed to fetch dynamically” errors, as was [Google News](https://news.google.com/search?q=500%20-%20Failed%20to%20fetch%20dynamically&hl=en-US&gl=US&ceid=US%3Aen) (which is especially bad for us, being a news site).

Our site is built with Nuxt 3 and deployed on Cloudflare Pages. Are you also using Cloudflare? I’m wondering if that’s a factor.

We tried various fixes (e.g. caching old assets for a year so they remained available), but the only thing that actually "worked" was disabling chunking:

  ```
vite: {
    build: {
      rollupOptions: {
        output: {
          inlineDynamicImports: true,
        },
      },
    },
```

Not ideal - we’ve definitely noticed a worse TTFP - but for now it’s the lesser of two evils while we work on cleaning up our search results. The date on the first news post (when searching for "500 - Failed to fetch dynamically") is now 1 week ago, which was roughly when we made the change, so there is definitely progress on that front at least.

This [Reddit thread](https://www.reddit.com/r/Nuxt/comments/1iu2f95/i_created_a_website_in_nuxt_and_hosted_it_but/) originally suggested the solution. As I understand it, the theory is that although Search Console renders the pages fine, Google’s crawler doesn’t always fetch all the chunks when there are too many, leading to the 500 being thrown. The evidence provided is somewhat anecdotal, but it does sound plausible.

Curious if anyone else has seen the same issue on Cloudflare, and what workarounds you’ve found.

Thanks!

**Comment by sewalsh** [comment]
> We tried various fixes (e.g. caching old assets for a year so they remained available), but the only thing that actually "worked" was disabling chunking:

I think there might be a misconception of how caching old assets works – do you mean that you've configured the HTTP headers to have a long cache time? If so, when you release subsequent builds, those 'old' cached files will still trigger a GET and will expect a `304` Not Modified HTTP response, but if you've purged the assets of the previous build(s), a `404` will be returned. Place all your build assets on a CDN/specific folder and don't purge the previous builds. Someone on [reddit](https://www.reddit.com/r/Nuxt/comments/1mx080x/need_help_with_chunkloaderror_in_a_highfrequency/) had a similar problem but it was their dev/test environment being updated constantly throughout the day.

**Comment by toniengelhardt** [comment]
> [@toniengelhardt](https://github.com/toniengelhardt)
> 
> We launched our new site at the end of July, and by mid-August [our SERPs](https://www.google.com/search?q=500+-+Failed+to+fetch+dynamically&sca_esv=bd4826c8ba41a3de&biw=2560&bih=1278&tbm=nws&sxsrf=AE3TifPmyo7BdRo_4gYR_h2zzQQP_W8RFw%3A1756241126924&ei=5hyuaJGaOMrXhbIP6tbakQY&ved=0ahUKEwiRpYqErKmPAxXKa0EAHWqrNmIQ4dUDCA4&uact=5&oq=500+-+Failed+to+fetch+dynamically&gs_lp=Egxnd3Mtd2l6LW5ld3MiITUwMCAtIEZhaWxlZCB0byBmZXRjaCBkeW5hbWljYWxseTIGEAAYFhgeMgYQABgWGB4yBhAAGBYYHjILEAAYgAQYhgMYigUyCxAAGIAEGIYDGIoFMgsQABiABBiGAxiKBTIIEAAYgAQYogQyBRAAGO8FMggQABiiBBiJBTIIEAAYgAQYogRIyAVQAFgAcAB4AJABAJgBWKABWKoBATG4AQPIAQCYAgGgAmKYAwCIBgGSBwExoAePB7IHATG4B2LCBwMyLTHIBwc&sclient=gws-wiz-news) were full of “500 – Failed to fetch dynamically” errors, as was [Google News](https://news.google.com/search?q=500%20-%20Failed%20to%20fetch%20dynamically&hl=en-US&gl=US&ceid=US%3Aen) (which is especially bad for us, being a news site).
> 
> Our site is built with Nuxt 3 and deployed on Cloudflare Pages. Are you also using Cloudflare? I’m wondering if that’s a factor.
> 
> We tried various fixes (e.g. caching old assets for a year so they remained available), but the only thing that actually "worked" was disabling chunking:
> 
> ```
> vite: {
>   build: {
>     rollupOptions: {
>       output: {
>         inlineDynamicImports: true,
>       },
>     },
>   },
> ```
> 
> Not ideal - we’ve definitely noticed a worse TTFP - but for now it’s the lesser of two evils while we work on cleaning up our search results. The date on the first news post (when searching for "500 - Failed to fetch dynamically") is now 1 week ago, which was roughly when we made the change, so there is definitely progress on that front at least.
> 
> This [Reddit thread](https://www.reddit.com/r/Nuxt/comments/1iu2f95/i_created_a_website_in_nuxt_and_hosted_it_but/) originally suggested the solution. As I understand it, the theory is that although Search Console renders the pages fine, Google’s crawler doesn’t always fetch all the chunks when there are too many, leading to the 500 being thrown. The evidence provided is somewhat anecdotal, but it does sound plausible.
> 
> Curious if anyone else has seen the same issue on Cloudflare, and what workarounds you’ve found.
> 
> Thanks!

Seems like a fix has been merged. Thanks @danielroe 🙏🏽

**Comment by angelorc** [comment]
So we just need to wait the new nuxt release? As usual @danielroe you are awesome.

**Comment by angelorc** [comment]
Solved in nuxt 4.1 🎉🎉🎉

**Comment by danielroe** [comment]
errors in google search of 500 - are suppressed (as of some releases ago), though you may need to get google to re-crawl those pages.

chunks will still encounter 404 errors, but with https://github.com/nuxt/nuxt/pull/33075 this should be less likely.

caching on your host remains the top cause of issues like this.

I'm marking the issue closed, however, as there's nothing specifically actionable for nuxt here.

**Comment by ahmadalfy** [comment]
@joeczucha We are using cloudflare too and faced the exact same problem. The solution we found was to purge the cache programatically after each deployment.

**Comment by initmn** [comment]
experienced the same thing in nuxt version 4.1.3, every time production build had to hard reload and clear browser cache to get rid of the error

**Comment by fikryrmdhna** [comment]
Can this method resolve it? I am using Nuxt version 3.12.1 @danielroe 

```
export default defineNuxtConfig({
  experimental: {
    entryImportMap: false
  },
})
```

**Comment by likt0r** [comment]
Same as  @initmn also on version 4.1.3.

**Comment by initmn** [comment]
@likt0r i fixed this issue. By removing the Lazy prefix component.

**Comment by menfukurou** [comment]
> @joeczucha We are using cloudflare too and faced the exact same problem. The solution we found was to purge the cache programatically after each deployment.

@ahmadalfy Which specific files did you purge from the cache?

**Comment by fikryrmdhna** [comment]
> [@likt0r](https://github.com/likt0r) i fixed this issue. By removing the Lazy prefix component.

do you use `experimental: { entryImportMap: false }` in your project? @initmn

**Comment by menfukurou** [comment]
Still having the same error.
I’m using Nuxt 3.19.2 deployed on Google App Engine.

This error occurs only when the Google bot crawls the site.

`'Failed to fetch dynamically imported module: https://xxxxxxx/_nuxt/DdmGmbLs.js', `

But the file does exist and can be retrieved by curl without any problem.

**Comment by initmn** [comment]
> > [@likt0r](https://github.com/likt0r) i fixed this issue. By removing the Lazy prefix component.
> 
> do you use `experimental: { entryImportMap: false }` in your project? [@initmn](https://github.com/initmn)

i don't

**Comment by initmn** [comment]
> Still having the same error. I’m using Nuxt 3.19.2 deployed on Google App Engine.
> 
> This error occurs only when the Google bot crawls the site.
> 
> `'Failed to fetch dynamically imported module: https://xxxxxxx/_nuxt/DdmGmbLs.js', `
> 
> But the file does exist and can be retrieved by curl without any problem.

try to update to version 3.19.3

**Comment by initmn** [comment]
> [@likt0r](https://github.com/likt0r) i fixed this issue. By removing the Lazy prefix component.

in cases nuxt version 3 and 4. just do this

**Comment by ahmadalfy** [comment]
> > [@joeczucha](https://github.com/joeczucha) We are using cloudflare too and faced the exact same problem. The solution we found was to purge the cache programatically after each deployment.
> 
> [@ahmadalfy](https://github.com/ahmadalfy) Which specific files did you purge from the cache?

Everything inside `nuxt` directory

**Comment by Eddymarais** [comment]
You can always compile the chunks in a new subfolder every time you build, this will ensure that the files don't get cached:

```
export default defineNuxtConfig({
  app: {
    baseURL: '/',
    buildAssetsDir: `/_nuxt/v${Math.floor(Date.now() / 1000).toString()}/`,
  },
```

You can also set this per environment: 
```
buildAssetsDir: (process.env.NODE_ENV === 'development') ? '/_nuxt/' : `/_nuxt/v${Math.floor(Date.now() / 1000).toString()}/`,
```

**Comment by Suniron** [comment]
I got a similar error with i18n, in a turborepo / bun monorepo:

<img width="1369" height="116" alt="Image" src="https://github.com/user-attachments/assets/e5a0e71b-a1b8-492f-b4b3-4c21a46e1003" />

I have a 500 error with these error:

```
Failed to fetch dynamically imported module: http://localhost:3000/_nuxt/@fs/home/user/projects/pro/my-project/apps/my-app/i18n/i18n.config.ts
```

To bypass it, I have to open a new tab in private mode

**Comment by bpcs-jan** [comment]
This is probably not a good solution and I don't recommend it but for us the error in the google search console was fixed by stripping the module preload links via this `stripModulePreload.ts` server plugin.

```
export default defineNitroPlugin((nitroApp) => {
	nitroApp.hooks.hook('render:html', (html) => {
		let totalKept = 0;

		html.head = html.head.map((chunk) => {
			if (typeof chunk !== 'string') return chunk;

			return chunk.replace(/<link[^>]*rel="modulepreload"[^>]*>/gi, (match) => {
				totalKept++;

				// Keep only the first 1 modulepreload
				if (totalKept <= 1) {
					return match;
				} else {
					return '';
				}
			});
		});
	});
});
```

And removing prefetch tags via this `stripPrefetchTags.ts` server plugin.

```
export default defineNitroPlugin((nitroApp) => {
	// Match modulepreload JS in /_nuxt/*.js|.mjs (any attr order/quotes/case, allow ?query)
	const RE_MODULEPRELOAD =
		/<link\b[^>]*\brel\s*=\s*(['"])modulepreload\1[^>]*\bas\s*=\s*(['"])script\2[^>]*\bhref\s*=\s*(['"])([^"']*\/_nuxt\/[^"']+\.(?:m?js)(?:\?[^"'>]*)?)\3[^>]*>/gi;

	// Match prefetch JS in /_nuxt/*.js|.mjs (same robustness)
	const RE_PREFETCH =
		/<link\b[^>]*\brel\s*=\s*(['"])prefetch\1[^>]*\bas\s*=\s*(['"])script\2[^>]*\bhref\s*=\s*(['"])([^"']*\/_nuxt\/[^"']+\.(?:m?js)(?:\?[^"'>]*)?)\3[^>]*>/gi;

	// Normalize href so absolute and relative compare equally
	const norm = (href: string) => {
		try {
			const u = new URL(href, 'https://dummy.base'); // base for relative URLs
			return u.pathname + u.search; // ignore origin
		} catch {
			return href;
		}
	};

	nitroApp.hooks.hook('render:html', (html) => {
		// 1) Collect all modulepreload hrefs (normalized)
		const moduleSet = new Set<string>();
		for (const chunk of html.head) {
			if (typeof chunk !== 'string') continue;
			let m: RegExpExecArray | null;
			RE_MODULEPRELOAD.lastIndex = 0;
			while ((m = RE_MODULEPRELOAD.exec(chunk))) {
				moduleSet.add(norm(m[4]));
			}
		}

		// Remove orphan prefetches
		let removedJsCount = 0;

		html.head = html.head.map((chunk) => {
			if (typeof chunk !== 'string') return chunk;

			// Remove orphan JS prefetches
			chunk = chunk.replace(RE_PREFETCH, (full, _q1, _q2, _q3, href) => {
				if (moduleSet.has(norm(href))) {
					return full;
				} else {
					removedJsCount++;
					return '';
				}
			});

			return chunk;
		});

		if (removedJsCount > 0) {
			console.log(`[Strip Prefetch] Removed ${removedJsCount} orphan JS`);
		}
	});
});
```


This also improved our lighthouse scores.

**Comment by gertjanjansen** [comment]
> Solved in nuxt 4.1 🎉🎉🎉

I believe this is not fixed in 4.1. I'm using Nuxt 4.1.3 with Vercel. After each new update Sentry reports a bunch of "Failed to fetch dynamically imported module ..." errors.

I tried `entryImportMap: false` but doesn't make a difference.

Which steps should I take to resolve this issue? Is this solved in the latest Nuxt version? Or should I enable Skew protection in Vercel? Even though the docs mention it's not ready for "zero configuration" yet (https://vercel.com/docs/skew-protection#supported-frameworks)

**Comment by YogeshwarDBarai** [comment]
Nuxt 3  - v3.20.1
Struggling to fix this error, I have made the file names consistent across every build, but now I am getting an error from the meta JSON 
like here 
`_nuxt/builds/meta/fd8630f0-1727-445f-a2d5-bdbf1bbae7ae.json`

I have configured Vite as mentioned below, and I have made the hash constant
`
onst fixedHash = 'somefixedrandomhash';	
```
vite: {
		build: {
		  rollupOptions: {
			output: {
				entryFileNames: `_nuxt/[name]-${fixedHash}.js`,
				chunkFileNames: `_nuxt/[name]-${fixedHash}.js`,
				assetFileNames: `_nuxt/[name]-${fixedHash}[extname]`
			}
		  }
		}
	},

```

Did anyone find a fix for this?
Google bot crawlers are mostly creating this issue for me and rarely for users, which is impacting SEO. The document returns 200, but the page content remains blank due to multiple failed resource issues. Google is caching the document HTML, and on my server, there is no document caching enabled. My JS files are being delivered via CDN. Every deployment invalidates the CDN cache .

**Comment by YogeshwarDBarai** [comment]
@danielroe 
Can we resolve this issue by retaining the old file name for a specific period of a month, so that if we encounter any old file names, they should point to the corresponding new file? 

I know this is a crazy idea, but yeah, I just wanted to mention it

**Comment by urbgimtam** [comment]
> You can always compile the chunks in a new subfolder every time you build, this will ensure that the files don't get cached:
> 
> ```
> export default defineNuxtConfig({
>   app: {
>     baseURL: '/',
>     buildAssetsDir: `/_nuxt/v${Math.floor(Date.now() / 1000).toString()}/`,
>   },
> ```
> 
> You can also set this per environment:
> 
> ```
> buildAssetsDir: (process.env.NODE_ENV === 'development') ? '/_nuxt/' : `/_nuxt/v${Math.floor(Date.now() / 1000).toString()}/`,
> ```

This idea of @Eddymarais sounds interesting to me, and would even minimize downtime while the server is building, if the previous folder would only be deleted after the build.

**Comment by Eddymarais** [comment]
> > You can always compile the chunks in a new subfolder every time you build, this will ensure that the files don't get cached:
> > ```
> > export default defineNuxtConfig({
> >   app: {
> >     baseURL: '/',
> >     buildAssetsDir: `/_nuxt/v${Math.floor(Date.now() / 1000).toString()}/`,
> >   },
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can also set this per environment:
> > ```
> > buildAssetsDir: (process.env.NODE_ENV === 'development') ? '/_nuxt/' : `/_nuxt/v${Math.floor(Date.now() / 1000).toString()}/`,
> > ```
> 
> This idea of [@Eddymarais](https://github.com/Eddymarais) sounds interesting to me, and would even minimize downtime while the server is building, if the previous folder would only be deleted after the build.

The previous folder should be deleted after every build and replaced with a new version.

**Comment by Bjornftw** [comment]
Still broken

<img width="1114" height="1186" alt="Image" src="https://github.com/user-attachments/assets/2a5f4565-68a3-4d1c-b9a4-a63fe1495f40" />

**Comment by KunMoe** [comment]
Even though this issue is closed, I want to add a **more precise root cause analysis**, because the same error is still affecting many Nuxt 3 users (including myself), and in many cases it is **not a Nuxt runtime bug**, but an interaction between **Nuxt’s chunk-splitting heuristics, dynamic imports, and CDN / firewall rules**.

I spent an entire day exhausting myself trying to troubleshoot this issue—building countless times and asking various LLMs countless questions. What I most want to say is this: **this issue should not be closed**. It still exists in the latest version, **Nuxt 4.2.2**, and **it seriously threatens what is claimed to be *Nuxt SEO***. I believe this issue deserves to be taken seriously.

In my case, with a Vue component, simply removing four or five lines of code made the problem
`Failed to fetch dynamically imported module "/_nuxt/<hash>.js"`
disappear. Adding those same four or five lines back caused the problem to reappear. This is not an occasional or random error—it is a **widespread** issue.

Below is a summary of the problem compiled by an LLM based on the various questions I asked throughout the day.


#### Symptom (same as this issue)

In production (but not dev):

```
TypeError: Failed to fetch dynamically imported module: /_nuxt/<hash>.js
```

This error:

* Does **not** happen for real users in normal browsers
* **Does** happen consistently in Google Search Console “URL Inspection / Live Test”
* Often cannot be reproduced locally
* Appears or disappears when a component gets slightly larger or more complex

---

## What actually triggers the error

### 1. Nuxt component auto-import + chunk splitting

In my case, the component is auto-registered via Nuxt’s component auto-import.

* When the component is small, Nuxt inlines it into the main client payload.
* When the component becomes *slightly* larger (e.g. adding one more `<select>`, a watcher, a store ref, or a small map), it crosses Nuxt/Vite’s internal heuristic threshold.
* At that point, Nuxt emits the component as a **separate lazy chunk** under `/_nuxt/<hash>.js`.

No logic bug is required — just a small increase in size/complexity.

---

### 2. Dynamic import during hydration

Once the component becomes its own chunk:

* SSR HTML still renders correctly
* During client hydration, Nuxt executes a dynamic `import()` to fetch that chunk

For real browsers:

* The request succeeds
* Hydration completes
* No visible error

---

### 3. Google’s renderer is *not* a normal browser

This is the key part that is usually missed.

Google Search Console / Googlebot uses **Web Rendering Service (WRS)**, which:

* Executes JavaScript
* **Fetches module chunks without typical browser headers**

  * No `Referer`
  * No `Sec-Fetch-Site`
  * No `Sec-Fetch-Mode`

---

### 4. CDN / firewall rules break `_nuxt/*.js` for bots

Many production setups (Cloudflare, WAFs, Nginx rules, “anti hotlink” logic) unintentionally do one of the following:

* Block requests to `/_nuxt/*.js` when `Referer` or `Sec-Fetch-*` headers are missing
* Treat such requests as suspicious and return:

  * an HTML fallback page
  * a soft 403
  * a rewritten response (still `200 OK`, but **HTML, not JS**)

For **real browsers**, everything works because those headers are present.

For **Googlebot**, the chunk request returns **HTML instead of JavaScript**.

---

### 5. Why the error message is misleading

From the JS engine’s point of view:

* `import("/_nuxt/<hash>.js")` *does* fetch something
* But the response body is **not a valid ES module**

At that point, the runtime throws:

```
TypeError: Failed to fetch dynamically imported module
```

Even though:

* HTTP status may be 200
* The file exists
* Manual `curl` or browser access works

This makes the problem look like a Nuxt/Vite bug, while the real cause is the **edge/CDN behavior difference for bots**.

---

## Why removing a small feature “fixes” it

In my case, removing a single feature (e.g. a “type” filter in a nav component):

* Made the component small enough to be inlined again
* Eliminated the dynamic import
* Prevented Googlebot from ever requesting `/_nuxt/<hash>.js`

So the page suddenly “works” again — without changing any server config.

This makes the bug extremely confusing and hard to diagnose.

---

## Why this issue keeps reappearing

* The error message points to Nuxt / Vite
* Real users cannot reproduce it
* `curl` tests usually pass
* The failure only happens in Google’s renderer
* The root cause lives outside Nuxt (CDN / WAF / edge rules)

As a result, many people still hit this problem even after this issue was closed.

---

## Practical fix (not in Nuxt code)

The real fix is **not** changing component logic, but ensuring that:

> **All requests to `/_nuxt/**` are allowed unconditionally for all User-Agents, regardless of headers.**

Specifically:

* Do NOT require `Referer`
* Do NOT require `Sec-Fetch-*`
* Do NOT apply bot challenges or hotlink protection to `/_nuxt/*.js`

Once this is done:

* Google URL Inspection stops reporting 500 / dynamic import errors
* No code changes are required
* SSR + hydration works as intended

---

## Conclusion

This issue looks like a Nuxt dynamic import bug, but in many real-world cases it is actually caused by:

> **Nuxt’s perfectly valid chunk splitting + Googlebot’s non-browser fetch behavior + overly strict CDN/firewall rules.**

I’m posting this here because even though the issue is closed, the underlying problem is still very real and still confusing a lot of users.

Hopefully this explanation helps others finally resolve it.

**Comment by bgarcias-cp** [comment]
I'm not sure why this issue was closed — it’s still happening in recent versions like Nuxt 4.1.3.

We are seeing many errors reported in Sentry, most of them coming from Firefox and Opera.

This happens even though we keep old build files available in Google Cloud Storage.

Our project is deployed on Google App Engine and we also use Cloudflare in front of it.

**Comment by jlemonz** [comment]
yes also still have this happening 

<img width="797" height="645" alt="Image" src="https://github.com/user-attachments/assets/f4e9ab97-d469-4596-be04-febb2c1c4521" />

**Comment by gertjanjansen** [comment]
Question: could it be that the `entryImportMap` and/or `appManifest` experimental config options are related to this issue?

**Comment by menfukurou** [comment]
The SEO issue in Google Search was resolved by applying the following `robots.txt` configuration:

User-agent: Googlebot
Disallow: /_nuxt/*.js

Note that client-side rendered content may not be correctly indexed, so this configuration should be applied based on the environment and use case.

**Comment by fruitoiz** [comment]
How to solve it? Or is there no solution yet?

**Comment by amaankm** [comment]
Anyone tried working with experimental `emitRouteChunkError` feature?
setting it to `automatic-immediate` immediately reloads the page on chunk error.
It internally calls uses the `reloadNuxtApp` which by default ignores the next reload call for 10s
```js
  experimental: {
    emitRouteChunkError: 'automatic-immediate',
  },
```

reference: 
https://nuxt.com/docs/4.x/api/utils/reload-nuxt-app 
https://nuxt.com/docs/4.x/guide/going-further/experimental-features#emitroutechunkerror

**Comment by dacaga** [comment]
Chiming in as a Nuxt 4 user. I went into this assuming the classic cause: a 404 from deploy timing while old tabs were still open. In my case it turned out to be a stale Cloudflare edge cache serving byte-different versions of the same content-hashed chunk (with vs. without the trailing //# sourceMappingURL= comment), which broke SRI integrity verification in the browser. Capturing more diagnostic info on these failures can help a lot!

**Comment by rahulkumarsingh73690** [comment]
This is a serious problem that mostly occurs when an existing tab is open after deployment. However, when I open a 404 chunk from the console, everything works. Why is this the case?

**Comment by harlan-zw** [comment]
I think this can stay closed as Nuxt is doing everything it itself can do and the path to solve this is documented.
One consideration which may be worth a follow-up is documenting that using caching on HTML pages will result in this issue happening much more frequently, especially when build-assets aren't served by CDNs.

I made the module https://github.com/harlan-zw/nuxt-skew-protection to help with this issue, if you're running into this issue give it a try.

> The SEO issue in Google Search was resolved by applying the following `robots.txt` configuration:
> 
> User-agent: Googlebot Disallow: /_nuxt/*.js
> 
> Note that client-side rendered content may not be correctly indexed, so this configuration should be applied based on the environment and use case.

Please don't do this; you'll block Google from indexing your site properly.

**Comment by Rolanddoda** [comment]
@harlan-zw this is an issue I am running too and closing this without a solution doesn't make sense I think.

**Comment by harlan-zw** [comment]
It would be better to create a new issue, this original issue is conflating a few issues into one. You can see the history of the issue and that several PRs have landed in Nuxt to resolve it from Nuxt's end.

The present issue, as far as I understand it, is outside of what the Nuxt framework can support without touching infrastructure. That being the issue of a browser or session hitting a HTML page (usually cached) that is requiring build assets that no longer exist. This happens between deployments when build assets change or when caching is misconfigured.

Nuxt provides primitives to help you handle new deployments through the [emitroutechunkerror hook](https://nuxt.com/docs/4.x/guide/going-further/experimental-features#emitroutechunkerror) and [experimental.checkOutdatedBuildInterval](https://nuxt.com/docs/4.x/guide/going-further/experimental-features). You can also configure a [cdnUrl](https://nuxt.com/docs/3.x/api/nuxt-config#cdnurl) to keep old build assets around longer.

Solving it at a higher level is not something that Nuxt core is attempting at the moment even though other frameworks do support it such as https://vercel.com/docs/skew-protection. This is because you start touching how your app is deployed and cached which Nuxt does not make judgements on (i.e Next.js skew protection only works on Vercel or [explicit setup for self-hosting](https://opennext.js.org/cloudflare/howtos/skew)). The [nuxt-skew-protection](https://github.com/harlan-zw/nuxt-skew-protection?rgh-link-date=2026-07-22T07%3A46%3A33.000Z) module is there to fill the gap until we explore this issue deeper.
